File: dan/Problems/Medium/LinkedList/148. Sort List/solution.py
# Definition for singly-linked list.
# class ListNode(object):
#     def __init__(self, x):
#         self.val = x
#         self.next = None

class Solution(object):
    #quick sort
    def sortList1(self, head):
        """
        :type head: ListNode
        :rtype: ListNode
        """
        return self.sort(head, None)
    
    def sort1(self, begin, end):
        if not begin or begin.next == end or begin == end:
            return begin
        i, j, x = begin, begin.next, begin.val
        while j is not end:
            while j is not end and j.val >= x:
                j = j.next
            if j is not end:
                i = i.next
                i.val, j.val = j.val, i.val
                j = j.next
        begin.val, i.val = i.val, begin.val
        self.sort(begin, i)
        self.sort(i.next, end)
        return begin
    #merge sort
    def merge(self, h1, h2):
        dummy = tail = ListNode(None)
        while h1 and h2:
            if h1.val < h2.val:
                tail.next, tail, h1 = h1, h1, h1.next
            else:
                tail.next, tail, h2 = h2, h2, h2.next
    
        tail.next = h1 or h2
        return dummy.next
    
    def sortList(self, head):
        if not head or not head.next:
            return head
    
        pre, slow, fast = None, head, head
        while fast and fast.next:
            pre, slow, fast = slow, slow.next, fast.next.next
        pre.next = None

        return self.merge(*map(self.sortList, (head, slow)))

File: dan/Problems/Medium/LinkedList/148. Sort List/solution.java
/**
 * Definition for singly-linked list.
 * public class ListNode {
 *     int val;
 *     ListNode next;
 *     ListNode(int x) { val = x; }
 * }
 */
class Solution {
    //Quicksort
    public ListNode sortList1(ListNode head) {
        return quickSort(head, null);
    }
    
    public ListNode quickSort(ListNode head, ListNode tail){
        if(head == tail || head.next == null)
            return head;
        ListNode i = head;
        ListNode j = head.next;
        int x = i.val;
        while(j != null){
            while(j != null && j.val >= x){
                j = j.next;
            }
            if(j != null){
                i = i.next;
                int t = i.val; i.val = j.val; j.val = t;
                j = j.next;
            }
        }
        int m = i.val; i.val = head.val; head.val = m;
        quickSort(head, i);
        quickSort(i.next, tail);
        return head;
    }
    
    //Merge sort
    public ListNode sortList(ListNode head){
        if(head == null || head.next == null)
            return head;
        //Find the middle node
        ListNode fast = head;
        ListNode slow = head;
        ListNode pre = null;
        while(fast != null && fast.next != null){
            pre = slow;
            fast = fast.next.next;
            slow = slow.next;
        }
        pre.next = null;
        return merge(sortList(head), sortList(slow));
    }
    
    public ListNode merge(ListNode h1, ListNode h2){
        if(h1 == null || h2 == null)
            return h1 == null? h2:h1;

        if(h1.val <= h2.val){
            h1.next = merge(h1.next, h2);
            return h1;
        }else{
            h2.next = merge(h1, h2.next);
            return h2;
        }
        
    }
}
